src/main/java/build/buildfarm/cas/GrpcCAS.java

```diff
@@ -0,0 +1,161 @@
+import com.google.bytestream.ByteStreamProto.ReadResponse;
+import com.google.bytestream.ByteStreamProto.QueryWriteStatusRequest;
+import com.google.bytestream.ByteStreamProto.QueryWriteStatusResponse;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.Iterators;
+import com.google.devtools.remoteexecution.v1test.ContentAddressableStorageGrpc;
+import com.google.devtools.remoteexecution.v1test.ContentAddressableStorageGrpc.ContentAddressableStorageBlockingStub;
+import com.google.devtools.remoteexecution.v1test.Digest;
+import com.google.protobuf.ByteString;
+import io.grpc.Channel;
+import io.grpc.StatusRuntimeException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+class GrpcCAS implements ContentAddressableStorage {
+  private final String instanceName;
+  private final Channel channel;
+  private final ByteStreamUploader uploader;
+  private final ConcurrentMap<Digest, List<Runnable>> digestsOnExpirations = new ConcurrentHashMap<>();
+
+  GrpcCAS(String instanceName, Channel channel, ByteStreamUploader uploader) {
+    this.instanceName = instanceName;
+    this.channel = channel;
+    this.uploader = uploader;
+  }
+
+  private final Supplier<ContentAddressableStorageBlockingStub> contentAddressableStorageBlockingStub =
+      Suppliers.memoize(
+          new Supplier<ContentAddressableStorageBlockingStub>() {
+            @Override
+            public ContentAddressableStorageBlockingStub get() {
+              return ContentAddressableStorageGrpc.newBlockingStub(channel);
+            }
+          });
+
+  private final Supplier<ByteStreamBlockingStub> bsBlockingStub =
+      Suppliers.memoize(
+          new Supplier<ByteStreamBlockingStub>() {
+            @Override
+            public ByteStreamBlockingStub get() {
+              return ByteStreamGrpc.newBlockingStub(channel);
+            }
+          });
+
+  private final Supplier<ByteStreamStub> bsStub =
+      Suppliers.memoize(
+          new Supplier<ByteStreamStub>() {
+            @Override
+            public ByteStreamStub get() {
+              return ByteStreamGrpc.newStub(channel);
+            }
+          });
+
+  public InputStream newStreamInput(String name) {
+    Iterator<ReadResponse> replies = bsBlockingStub
+        .get()
+        .read(ReadRequest.newBuilder().setResourceName(name).build());
+    return new ByteStringIteratorInputStream(Iterators.transform(replies, (reply) -> reply.getData()));
+  }
+
+  private String getBlobName(Digest digest) {
+    return String.format(
+        "%sblobs/%s",
+        instanceName.isEmpty() ? "" : (instanceName + "/"),
+        DigestUtil.toString(digest));
+  }
+
+  @Override
+  public boolean contains(Digest digest) {
```

Comment:
I don't know what getMissing is, it doesn't appear in the bytestream API. Do you mean `findMissingBlobs` from the cas? The presumption for me was that we could use a bytestream only implementation for this cas, instead of requiring the (less common) ContentAddressableStorage implementation from remote_execution.

I'm expiring it so that on any access (since I don't have a callback mechanism for bytestream, though this might be a good case for a watcher), either get or contains, we trigger the onExpiration for a key. poor man's event system.

Note the comparison to the committed size: only the completed upload with committed size maching the digest size is considered 'contained' - we cannot return true for contains when the committed size has not reached the digest-indicated size. This check was actually missing from `get`, I will add it.